### DISCUSSION TITLE
How can I use local firmware to upgrade my device

### DISCUSSION THREAD

**Discussion by Aogu181** [Koenkk/zigbee2mqtt#24339] [General]
I know Z2M has a local OTA index, but I don't know how to use it. Or who can help me upgrade my device using local firmware

**Comment by usma0118** [comment]
look here:
https://www.zigbee2mqtt.io/guide/usage/ota_updates.html#local-ota-index-and-firmware-files

and https://smarthomescene.com/guides/how-to-setup-zigbee2mqtt-local-ota-index/

**Reply by Aogu181** [reply]
I have tried.  Log as follows: 
z2m: Update of '008P' failed (Cannot destructure property 'data' of '(intermediate value)' as it is undefined.)

**Reply by JohnRo79** [reply]
> I have tried. Log as follows: z2m: Update of '008P' failed (Cannot destructure property 'data' of '(intermediate value)' as it is undefined.)

i've used this this site   https://darkxst.github.io/cp-test/

but without knowing what device you've got, i don't know...
